### ISSUE TITLE
in column info for "TypeBit", "Length" should not be multiplied by 4

### ISSUE THREAD

**Issue by funow** [pingcap/tidb#5405]
Please answer these questions before submitting your issue. Thanks!

1. What did you do?
I have failed to query a table including BIT(1) field, using statement:
`select * from CACHE_DICT  t  order by dict_id desc limit 0,10`
The table is created by:
```
CREATE TABLE `cache_dict` (
  `DICT_ID` int(11) NOT NULL AUTO_INCREMENT,
  `DICT_CAT` char(50) DEFAULT NULL COMMENT '分类',
  `DICT_UUID` varchar(2000) DEFAULT NULL COMMENT '识别码',
  `PARENT_DICT_ID` int(11) DEFAULT NULL COMMENT '上级字典ID',
  `DICT_SEQ` int(11) DEFAULT NULL COMMENT '字典序号',
  `DEF_DICT_NAME` char(100) DEFAULT NULL COMMENT '字典名',
  `DEFUNT_IND` bit(1) DEFAULT NULL COMMENT '有效标志',
  `DEF_IND` bit(1) DEFAULT NULL COMMENT '默认值',
  `REMARK` varchar(2000) DEFAULT NULL COMMENT '备注',
  PRIMARY KEY (`DICT_ID`)
)
```

2. What did you expect to see?

3. What did you see instead?
```
2017/12/14 16:40:39.538 conn.go:429: [warning] [142] dispatch error:
id:142, addr:192.168.14.209:9702 status:2, collation:utf8_general_ci, user:root
"select * from CACHE_DICT  t  order by dict_id desc limit 0,10"
/home/jenkins/workspace/build_tidb_master/go/src/github.com/pingcap/tidb/server/util.go:346: [server:4]invalid type 16
/home/jenkins/workspace/build_tidb_master/go/src/github.com/pingcap/tidb/server/conn.go:832: 
/home/jenkins/workspace/build_tidb_master/go/src/github.com/pingcap/tidb/server/conn.go:756: 
2017/12/14 16:40:39.539 terror.go:281: [warning] Unknown error class: server code: 4
```

4. What version of TiDB are you using (`tidb-server -V`)?
Both 1.0.3 and 1.1.0-alpha

**Comment by jackysp** [comment]
PTAL @winoros

**Comment by shenli** [comment]
@funow I can not reproduce this. Could you show us the output of the `select tidb_version();` statement? We need the exact githash.
```sql
mysql> CREATE TABLE `cache_dict` (
    ->   `DICT_ID` int(11) NOT NULL AUTO_INCREMENT,
    ->   `DICT_CAT` char(50) DEFAULT NULL COMMENT '分类',
    ->   `DICT_UUID` varchar(2000) DEFAULT NULL COMMENT '识别码',
    ->   `PARENT_DICT_ID` int(11) DEFAULT NULL COMMENT '上级字典ID',
    ->   `DICT_SEQ` int(11) DEFAULT NULL COMMENT '字典序号',
    ->   `DEF_DICT_NAME` char(100) DEFAULT NULL COMMENT '字典名',
    ->   `DEFUNT_IND` bit(1) DEFAULT NULL COMMENT '有效标志',
    ->   `DEF_IND` bit(1) DEFAULT NULL COMMENT '默认值',
    ->   `REMARK` varchar(2000) DEFAULT NULL COMMENT '备注',
    ->   PRIMARY KEY (`DICT_ID`)
    -> );
Query OK, 0 rows affected (0.01 sec)

mysql> select * from CACHE_DICT t order by dict_id desc limit 0,10;
Empty set (0.10 sec)

mysql> insert into CACHE_DICT values (1, "cat", "uuid", 10, 10, "def_dict_name", 1, 1, "remark");
Query OK, 1 row affected (0.00 sec)

mysql> select * from CACHE_DICT t order by dict_id desc limit 0,10;
+---------+----------+-----------+----------------+----------+---------------+------------+---------+--------+
| DICT_ID | DICT_CAT | DICT_UUID | PARENT_DICT_ID | DICT_SEQ | DEF_DICT_NAME | DEFUNT_IND | DEF_IND | REMARK |
+---------+----------+-----------+----------------+----------+---------------+------------+---------+--------+
|       1 | cat      | uuid      |             10 |       10 | def_dict_name |           |        | remark |
+---------+----------+-----------+----------------+----------+---------------+------------+---------+--------+
1 row in set (0.01 sec)

mysql> select tidb_version();
+--------------------------------------------------------------------------------------------------------------------------------------------------------------+
| tidb_version()                                                                                                                                               |
+--------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Release Version: v1.1.0-alpha-267-g37eb5403
Git Commit Hash: 37eb54038609cf414e7f9a797981c0984d6d98e5
Git Branch: master
UTC Build Time: 2017-12-14 09:19:04 |
+--------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```

**Comment by funow** [comment]
version
Release Version: v1.1.0-alpha-71-gec9d1f0
Git Commit Hash: ec9d1f0cc6cd6045b56f3d3f4be32653230238b8
Git Branch: master
UTC Build Time: 2017-11-12 04:08:24

**Comment by shenli** [comment]
This is fixed. Please try the latest master.
Fixed in https://github.com/pingcap/tidb/pull/5066

**Comment by funow** [comment]
My INSERT statement is a little different from yours. I used statement like
`INSERT INTO `cache_dict` VALUES (25, 'INSTALL_TYPE', 'syncdata', 0, 0, '缓存自维护', b'1', b'1', '');`
and
`INSERT INTO `cache_dict` VALUES ('2', 'installType_projectName_map', 'ctg-cache-access', '0', '1', 'access', '\0', '\0', '安装类型和项目名称映射');`
But the INSERT statement is successful.

**Comment by funow** [comment]
I tried version 1.0.4, and the query is successful.
But when we trying to cast the BIT(1) field to **boolean** using JDBC, it came out with a ClassCastException
`Exception in thread "main" java.lang.ClassCastException: [B cannot be cast to java.lang.Boolean`

We tried ResultSet.getObject().toString(), and what TiDB returned is something like `[B@15327b79`, but what MYSQL returned is just `false` or `true`.

**Comment by shenli** [comment]
@funow May I have your SQL or code to reproduce this issue?
@XuHuaiyu @zz-jason PTAL

**Comment by funow** [comment]
This is our test code, it works with MYSQL, but it throws exception with TiDB.
```
public static String sql = "select defunt_ind from cache_dict where dict_id=2";
public void test() {
		try {
			ps = conn.prepareStatement(sql);
			rs = ps.executeQuery();
			while(rs.next()) {
				Object o = rs.getObject(1);
				System.out.println(o.toString());
				System.out.println((boolean)o);
			}
		} catch (SQLException e) {
			e.printStackTrace();
		}finally {
			try {
				rs.close();
				ps.close();
				conn.close();
			} catch (SQLException e) {
				e.printStackTrace();
			}
		}
	}
```

**Comment by zz-jason** [comment]
@funow Thanks for your feed back! 

For the first question:`select defunt_ind from cache_dict where dict_id=2` failed with "ERROR 1105 (HY000): invalid type 16", this was fixed in https://github.com/pingcap/tidb/pull/5066

For the second question, witch is:
> But when we trying to cast the BIT(1) field to boolean using JDBC, it came out with a ClassCastException
> `Exception in thread "main" java.lang.ClassCastException: [B cannot be cast to java.lang.Boolean`

Can you provide us the query sql or code so we can reproduce this problem ?

**Comment by shenli** [comment]
@zz-jason The code is in the comment above.

**Comment by zz-jason** [comment]
@shenli @funow for the second issue, this is due to the incorrect `Length` for type `bit(1)`:

in TiDB the `Length` is `4`
```sql
TiDB(localhost) > select DEFUNT_IND from cache_dict;
Field   1:  `DEFUNT_IND`
Catalog:    `def`
Database:   `test`
Table:      `cache_dict`
Org_table:  `cache_dict`
Type:       BIT
Collation:  binary (63)
Length:     4
Max_length: 1
Decimals:   0
Flags:      UNSIGNED


+------------+
| DEFUNT_IND |
+------------+
|           |
+------------+
1 row in set (0.00 sec)
```

but it is `1` in MySQL, witch is the same as the create table statement:
```sql
MySQL(localhost) > select DEFUNT_IND from cache_dict;
Field   1:  `DEFUNT_IND`
Catalog:    `def`
Database:   `test`
Table:      `cache_dict`
Org_table:  `cache_dict`
Type:       BIT
Collation:  binary (63)
Length:     1
Max_length: 1
Decimals:   0
Flags:      UNSIGNED


+------------+
| DEFUNT_IND |
+------------+
|           |
+------------+
1 row in set (0.00 sec)
```

**Comment by zz-jason** [comment]
```txt
323 func convertColumnInfo(fld *ast.ResultField) (ci *ColumnInfo) {
324     ci = new(ColumnInfo)
325     ci.Name = fld.ColumnAsName.O
326     ci.OrgName = fld.Column.Name.O
327     ci.Table = fld.TableAsName.O
328     if fld.Table != nil {
329         ci.OrgTable = fld.Table.Name.O
330     }
331     ci.Schema = fld.DBName.O
332     ci.Flag = uint16(fld.Column.Flag)
333     ci.Charset = uint16(mysql.CharsetIDs[fld.Column.Charset])
334     if fld.Column.Flen == types.UnspecifiedLength {
335         ci.ColumnLength = 0
336     } else {
337         ci.ColumnLength = uint32(fld.Column.Flen)
338     }
339     if fld.Column.Tp == mysql.TypeNewDecimal {
340         // Consider the negative sign.
341         ci.ColumnLength++
342         if fld.Column.Decimal > types.DefaultFsp {
343             // Consider the decimal point.
344             ci.ColumnLength++
345         }
346     } else {
347         // Fix issue #4540.
348         // The flen is a hint, not a precise value, so most client will not use the value.
349         // But we found in race MySQL client, like Navicat for MySQL(version before 12) will truncate
350         // the `show create table` result. To fix this case, we must use a large enough flen to prevent
351         // the truncation, in MySQL, it will multiply bytes length by a multiple based on character set.
352         // For examples:
353         // * latin, the multiple is 1
354         // * gb2312, the multiple is 2
355         // * Utf-8, the multiple is 3
356         // * utf8mb4, the multiple is 4
357         // So the large enough multiple is 4 in here.
358         ci.ColumnLength = ci.ColumnLength * mysql.MaxBytesOfCharacter
359     }
```

in line 358, we should not simply multiply `mysql.MaxBytesOfCharacter`
